File: packages/deepfreeze-kibana/server/routes/status.ts
import { schema } from '@kbn/config-schema';
import { IRouter } from '@kbn/core/server';
import { DeepfreezeClient } from '../lib/deepfreeze_client';
import { API_BASE } from '../../common';

export function registerStatusRoutes(router: IRouter, client: DeepfreezeClient) {
  // Full status
  router.get(
    {
      path: `${API_BASE}/status`,
      validate: {
        query: schema.object({
          force_refresh: schema.boolean({ defaultValue: false }),
        }),
      },
    },
    async (_context, request, response) => {
      try {
        const result = await client.get('/api/status', {
          force_refresh: request.query.force_refresh,
        });
        return response.ok({ body: result });
      } catch (error: any) {
        return response.customError({
          statusCode: error.statusCode || 502,
          body: { message: error.message },
        });
      }
    },
  );

  // Sub-resources
  for (const sub of ['cluster', 'repositories', 'thaw-requests', 'buckets', 'ilm-policies']) {
    router.get(
      { path: `${API_BASE}/status/${sub}`, validate: false },
      async (_context, _request, response) => {
        try {
          const result = await client.get(`/api/status/${sub}`);
          return response.ok({ body: result });
        } catch (error: any) {
          return response.customError({
            statusCode: error.statusCode || 502,
            body: { message: error.message },
          });
        }
      },
    );
  }

  // Restore progress
  router.get(
    {
      path: `${API_BASE}/thaw-requests/{requestId}/restore-progress`,
      validate: {
        params: schema.object({ requestId: schema.string() }),
      },
    },
    async (_context, request, response) => {
      try {
        const result = await client.get(
          `/api/thaw-requests/${request.params.requestId}/restore-progress`,
        );
        return response.ok({ body: result });
      } catch (error: any) {
        return response.customError({
          statusCode: error.statusCode || 502,
          body: { message: error.message },
        });
      }
    },
  );

  // History
  router.get(
    {
      path: `${API_BASE}/history`,
      validate: {
        query: schema.object({
          limit: schema.number({ defaultValue: 25 }),
        }),
      },
    },
    async (_context, request, response) => {
      try {
        const result = await client.get('/api/history', { limit: request.query.limit });
        return response.ok({ body: result });
      } catch (error: any) {
        return response.customError({
          statusCode: error.statusCode || 502,
          body: { message: error.message },
        });
      }
    },
  );

  // Audit log
  router.get(
    {
      path: `${API_BASE}/audit`,
      validate: {
        query: schema.object({
          limit: schema.number({ defaultValue: 50 }),
          action: schema.maybe(schema.string()),
        }),
      },
    },
    async (_context, request, response) => {
      try {
        const params: Record<string, string | number> = { limit: request.query.limit };
        if (request.query.action) {
          params.action = request.query.action;
        }
        const result = await client.get('/api/audit', params);
        return response.ok({ body: result });
      } catch (error: any) {
        return response.customError({
          statusCode: error.statusCode || 502,
          body: { message: error.message },
        });
      }
    },
  );
}

File: packages/deepfreeze-server/deepfreeze_server/api/status.py
"""Status and read-only API routes."""

import asyncio

from fastapi import APIRouter, Depends, Query

from ..orchestration.orchestrator import DeepfreezeOrchestrator
from .deps import get_orchestrator

router = APIRouter()


@router.get("/status")
async def get_status(
    force_refresh: bool = False,
    orch: DeepfreezeOrchestrator = Depends(get_orchestrator),
):
    """Get full system status (cluster, repos, thaw requests, buckets, ILM)."""
    status = await orch.status_cache.get_status(force_refresh=force_refresh)
    return status.model_dump()


@router.get("/status/cluster")
async def get_cluster_health(
    orch: DeepfreezeOrchestrator = Depends(get_orchestrator),
):
    """Get cluster health only."""
    status = await orch.status_cache.get_status()
    return status.cluster.model_dump()


@router.get("/status/repositories")
async def get_repositories(
    orch: DeepfreezeOrchestrator = Depends(get_orchestrator),
):
    """Get all repositories."""
    status = await orch.status_cache.get_status()
    return {"repositories": status.repositories}


@router.get("/status/thaw-requests")
async def get_thaw_requests(
    orch: DeepfreezeOrchestrator = Depends(get_orchestrator),
):
    """Get all thaw requests."""
    status = await orch.status_cache.get_status()
    return {"thaw_requests": status.thaw_requests}


@router.get("/status/buckets")
async def get_buckets(
    orch: DeepfreezeOrchestrator = Depends(get_orchestrator),
):
    """Get all buckets."""
    status = await orch.status_cache.get_status()
    return {"buckets": status.buckets}


@router.get("/status/ilm-policies")
async def get_ilm_policies(
    orch: DeepfreezeOrchestrator = Depends(get_orchestrator),
):
    """Get all ILM policies."""
    status = await orch.status_cache.get_status()
    return {"ilm_policies": status.ilm_policies}


@router.get("/history")
async def get_action_history(
    limit: int = Query(default=25, ge=1, le=500),
    orch: DeepfreezeOrchestrator = Depends(get_orchestrator),
):
    """Get recent action history."""
    loop = asyncio.get_running_loop()
    history = await loop.run_in_executor(
        None, orch.status_cache.get_action_history, limit
    )
    return {"history": [h.model_dump() for h in history]}


@router.get("/thaw-requests/{request_id}/restore-progress")
async def get_restore_progress(
    request_id: str,
    orch: DeepfreezeOrchestrator = Depends(get_orchestrator),
):
    """Get S3 restore progress for each repo in a thaw request."""
    progress = await orch.get_thaw_restore_progress(request_id)
    return {"request_id": request_id, "repos": progress}


@router.get("/audit")
async def get_audit_log(
    limit: int = Query(default=50, ge=1, le=500),
    action: str | None = None,
    orch: DeepfreezeOrchestrator = Depends(get_orchestrator),
):
    """Get detailed audit log entries from the ES audit index."""
    audit = orch._audit
    if not audit:
        return {"entries": [], "source": "unavailable"}
    loop = asyncio.get_running_loop()
    entries = await loop.run_in_executor(
        None, lambda: audit.get_recent_entries(limit=limit, action_filter=action)
    )
    return {"entries": entries, "source": "elasticsearch"}
